File: CW/Lista_1/Z5.scala
def palindrome[A](xs: List[A]): Boolean = {
    def palindromeRec(l: Int, r: Int): Boolean = {
        if(l >= r) true
        else if (xs(l) != xs(r)) false
        else palindromeRec(l + 1, r - 1)
    }

    palindromeRec(0, xs.length - 1)
}

@main def main() = {
    val list = List(1, 2, 3, 2, 1, 5)
    println(palindrome(list))
}

File: CW/Lista_1/Z5.ml
let palindrome (xs: 'a list): bool =
  let rec reverse (xs: 'a list): 'a list =
    if xs = [] then
      []
    else
      let head = List.hd xs in
      let tail = List.tl xs in
      (reverse tail) @ [head]
  in
  
  xs = reverse xs

let lst = [1;3;1];;
assert (palindrome lst = true);;